File: fb_encoder.cpp
#include <fstream>
#include <iostream>
#include <string>
#include <vector>
#include "client_generated.h"

using namespace CrossParser;

int main(int argc, char **argv) {
    if (argc < 2) {
        std::cerr << "Format: ./fb_encoder <path-to-fb_bytes.bin>\n";
        return 1;
    }

    std::string file_path(argv[1]);
    file_path += "fb_bytes.bin";
    std::ofstream outfile(file_path, std::ios::out | std::ios::binary);
    if (!outfile.is_open()) {
        std::cerr << "Issue with file opening, please check\n";
        return 2;
    }

    flatbuffers::FlatBufferBuilder builder;
    
    auto client_person_name = builder.CreateString("Ram");
    double client_person_age = 21;
    double client_person_weight = 76.5;
    auto client_person_gender = builder.CreateString("Male");

    auto client_group_name = builder.CreateString("FightClub");
    double client_group_avg_age = 24.5;
    double client_group_avg_weight = 66;
    std::vector<std::string> tmp_vec_names{"Ram", "Shyam", "Raghuveer"};
    auto client_group_vec_names = builder.CreateVectorOfStrings(tmp_vec_names);

    auto person = CreatePerson(builder, client_person_name, client_person_age, client_person_weight, client_person_gender);
    auto group = CreateGroup(builder, client_group_name, client_group_avg_age, client_group_avg_weight, client_group_vec_names);

    auto client_person = CreateClient(builder, ClientUnion_Person, person.Union());
    auto client_group = CreateClient(builder, ClientUnion_Group, group.Union());

    std::vector<flatbuffers::Offset<Client>> tmp_vec_clients{client_person, client_group};
    auto clients_vector = builder.CreateVector(tmp_vec_clients);
    auto root_client_vector = CreateClientVec(builder, clients_vector);

    builder.Finish(root_client_vector);
    uint8_t *buf = builder.GetBufferPointer();
    int size = builder.GetSize();
    outfile.write(reinterpret_cast<const char*>(buf), size);
    outfile.close();
    return 0;
}

File: fb_encoder.py
import os, sys
sys.path.append(os.path.join(os.path.dirname(__file__), 'flatbuffers/python/'))

import flatbuffers

import CrossParser.Person
import CrossParser.Group
import CrossParser.ClientUnion 
import CrossParser.Client
import CrossParser.ClientVec

if __name__ == '__main__':
    if len(sys.argv) < 2:
        print("Format: fb_encoder.py <path-to-fb_bytes.bin>", file=sys.stderr)
        exit(1)

    file_path = sys.argv[1]
    file_path += "fb_bytes.bin"

    with open(file_path, "wb") as outfile:

        builder = flatbuffers.Builder(0)

        client_person_name = builder.CreateString("Ram")
        client_person_gender = builder.CreateString("Male")
        CrossParser.Person.Start(builder)
        CrossParser.Person.AddName(builder, client_person_name)
        CrossParser.Person.AddAge(builder, 21)
        CrossParser.Person.AddWeight(builder, 76.5)
        CrossParser.Person.AddGender(builder, client_person_gender)
        person = CrossParser.Person.End(builder)

        client_group_name = builder.CreateString("FightClub")
        string_offset_list = []
        string_offset_list.append(builder.CreateString("Raghuveer"))
        string_offset_list.append(builder.CreateString("Shyam"))
        string_offset_list.append(builder.CreateString("Ram"))
        CrossParser.Group.StartPersonNamesVector(builder, 3)
        for string_offset in string_offset_list:
            builder.PrependUOffsetTRelative(string_offset)
        client_group_vec_names = builder.EndVector() 
        CrossParser.Group.Start(builder)
        CrossParser.Group.AddGroupName(builder, client_group_name)
        CrossParser.Group.AddAvgAge(builder, 24.5)
        CrossParser.Group.AddAvgWeight(builder, 66)
        CrossParser.Group.AddPersonNames(builder, client_group_vec_names)
        group = CrossParser.Group.End(builder)

        CrossParser.Client.Start(builder)
        CrossParser.Client.AddClientElemType(builder, CrossParser.ClientUnion.ClientUnion().Person)
        CrossParser.Client.AddClientElem(builder, person)
        client_person = CrossParser.Client.End(builder)

        CrossParser.Client.Start(builder)
        CrossParser.Client.AddClientElemType(builder, CrossParser.ClientUnion.ClientUnion().Group) 
        CrossParser.Client.AddClientElem(builder, group)
        client_group = CrossParser.Client.End(builder)

        CrossParser.ClientVec.StartClientVecVector(builder, 2)
        for offset_val in [client_group, client_person]:
            builder.PrependUOffsetTRelative(offset_val)
        clients_vector = builder.EndVector()
        CrossParser.ClientVec.Start(builder)
        CrossParser.ClientVec.AddClientVec(builder, clients_vector)
        root_client_vector = CrossParser.ClientVec.End(builder)

        builder.Finish(root_client_vector)
        buf = builder.Output()
        outfile.write(buf)
